kong/runloop/balancer/ewma.lua

```diff
@@ -0,0 +1,255 @@
+-- of existing endpoints.
+local function calculate_slow_start_ewma(self)
+    local total_ewma = 0
+    self.address_count = 0
+  
+    for _, target in ipairs(self.balancer.targets) do
+        for _, address in ipairs(target.addresses) do
+            if address.available then
+              local ewma = self.ewma[address] or 0
+              self.address_count = self.address_count + 1
+              total_ewma = total_ewma + ewma
+            end
+        end
+    end
+  
+    if self.address_count == 0 then
+      ngx_log(ngx_DEBUG, "no ewma value exists for the endpoints")
+      return nil
+    end
+  
+    return total_ewma / self.address_count
+end
+
+
+function ewma:afterHostUpdate()
+  local new_addresses = {}
+  
+  for _, target in ipairs(self.balancer.targets) do
+    for _, address in ipairs(target.addresses) do
+      if address.available then
+        new_addresses[address] = true
+      end
+    end
+  end
+
+  for address, _ in pairs(self.ewma) do
+    if not new_addresses[address] then
+      self.ewma[address] = nil
+      self.ewma_last_touched_at[address] = nil
+    end
+  end
+  
+  local slow_start_ewma = calculate_slow_start_ewma(self)
+  if slow_start_ewma ~= nil then
+    local now = ngx_now()
+    for address, _ in pairs(new_addresses) do
+      self.ewma[address] = slow_start_ewma
+      self.ewma_last_touched_at[address] = now
+    end
+  end
+end
+
+
+local function get_or_update_ewma(self, address, rtt, update)
+  local ewma = self.ewma[address] or 0
+  local now = ngx_now()
+  local last_touched_at = self.ewma_last_touched_at[address] or 0
+  ewma = decay_ewma(ewma, last_touched_at, rtt, now)
+  if not update then
+    return ewma
+  end
+
+  self.ewma_last_touched_at[address] = now
+  self.ewma[address] = ewma
+  return ewma
+end
+
+
+function ewma:afterBalance(ctx, handle)
+  local response_time = ngx.var.upstream_response_time or 0
+  local connect_time = ngx.var.upstream_connect_time or 0
+  local rtt = connect_time + response_time
+  local upstream = ngx.var.upstream_addr
+  local address = handle.address
+
+  if not upstream then
+      return nil, "no upstream addr found"
+  end
+  ngx_log(ngx_DEBUG, "ewma after balancer rtt: ", rtt)
+  return get_or_update_ewma(self, address, rtt, true)
+end
+
+
+-- implementation similar to https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle
+-- or https://en.wikipedia.org/wiki/Random_permutation
+-- loop from 1 .. k
+-- pick a random value r from the remaining set of unpicked values (i .. n)
+-- swap the value at position i with the value at position r
+local function shuffle_address(address, k)
+  for i=1, k do
+    local rand_index = math.random(i,#address)
+    address[i], address[rand_index] = address[rand_index], address[i]
+  end
+  -- peers[1 .. k] will now contain a randomly selected k from #peers
+end
+
+local function pick_and_score(self, address, k)
+  shuffle_address(address, k)
```

Comment:
Why are the addresses shuffled before getting the lowest score?

Reply:
consider this situation, when the ewma balancer just starts, all ewma score is 0, we need to select one node randomly. not the first node.

Reply:
Could you get the same result with something along the line of: `local lowest_score_index = math.random(1, k)`? We would avoid a `for` loop on every `getPeer()` call.